CDAP-13096 refactor program lifecycle to trigger off TMS

albertshau opened the pull request:
Refactoring so that program execution will be triggered off events
on TMS in preparation for additional cluster lifecycle states.
Refactoring the ProgramLifecycleService to remove usage of
ProgramController when possible, as starting a run will first need
to provision the runtime before the ProgramController can be
used.

Added a 'pending' program run status to indicate the actual
program has not been started. This corresponds to the provisioning
cluster state. Also adding a 'starting' program state to the
status rest endpoint, as a 'running' status when a cluster is
being provisioned is quite misleading, and breaks a bunch of
unit tests that assume a 'running' program status means everything,
like service instances, are actually up and running.
Along with this change, fixing unit tests to define isRunning
as anything that is not stopped.

albertshau committed:
CDAP-13096 refactor program lifecycle to trigger off TMS

Refactoring so that program execution will be triggered off events
on TMS in preparation for additional cluster lifecycle states.
Refactoring the ProgramLifecycleService to remove usage of
ProgramController when possible, as starting a run will first need
to provision the runtime before the ProgramController can be
used.

Added a 'pending' program run status to indicate the actual
program has not been started. This corresponds to the provisioning
cluster state. Also adding a 'starting' program state to the
status rest endpoint, as a 'running' status when a cluster is
being provisioned is quite misleading, and breaks a bunch of
unit tests that assume a 'running' program status means everything,
like service instances, are actually up and running.
Along with this change, fixing unit tests to define isRunning
as anything that is not stopped.

albertshau commented:
https://builds.cask.co/browse/CDAP-DUT6334

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/ScheduleTaskRunner.java:
```diff
@@ -130,50 +119,14 @@ private TriggeringScheduleInfo getTriggeringScheduleInfo(Job job) {
-      runtimeInfo = lifecycleService.startInternal(id, sysArgs, userArgs, false);
+      ProgramOptions programOptions = new SimpleProgramOptions(id, new BasicArguments(sysArgs),
+                                                               new BasicArguments(userArgs), false);
+      provisionerNotifier.provisioning(id.run(RunIds.generate()), programOptions);
```
Why interact with `ProvisionerNotifier` directly? I would expect the AppLifecycleService is the only integration point for other components.

chtyim replied:
The reason for that is because there would be new lifecycle methods to be run before provision and after deprovisioning. Exposing the internal mechanism outside of the AppLifecycleService breaks the abstraction boundary.

albertshau replied:
I was running into issues instantiating the ProgramLifecycleService in unit test because there is a cyclic dependency for guice injection with ProgramLifecycleService -> CoreSchedulerService -> ConstraintCheckerService -> ProgramLifecycleService. Though it must be working somehow in sdk and distributed, so I'll take another look at the guice modules and see what can be done.

chtyim replied:
Seems like the circular dependency is a wrong construct. The `ProgramLifecycleService` should NOT take a `Scheduler`, since those functions (e.g. getScheduleStatus, suspendResumeSchedule, etc.) belongs to the Scheduler service, not program lifecycle.

albertshau replied:
agreed, I may open a separate PR to do this refactoring first

albertshau committed:
replace waitForStatus(true) calls in unit tests

This method is almost always subjected to race conditions, where
a program is starting but not yet running, and other parts of the
unit test expect it to be running. The existing race conditions
are made much more likely due to the additional provisioning
lifecycle stages, which causes programs to get to the running
state a little slower now.

albertshau commented on cdap-app-fabric-tests/src/test/java/co/cask/cdap/internal/app/services/http/handlers/WorkflowStatsSLAHttpHandlerTest.java:
```diff
@@ -86,7 +86,8 @@ private void setStartAndRunning(final ProgramId id, final String pid, final long
                           AppFabricTestHelper.createSourceId(++sourceId));
     store.setProvisioned(id.run(pid), 0, AppFabricTestHelper.createSourceId(++sourceId));
-    store.setStart(id.run(pid), null, systemArgs, AppFabricTestHelper.createSourceId(++sourceId));
+    store.setStart(id.run(pid), startTime, null, runtimeArgs, systemArgs,
```
I had removed start time and runtime args out in an earlier refactoring, but realized we still need them for programs that are started by preview and programs started within a workflow. So these types of changes are just adding back what used to be there.

albertshau commented:
core changes are in ProgramLifecycleService, ProgramNotificationSubscriberService, and ProvisionerNotificationSubscriberService. There are a lot of test changes due to existing race conditions made much worse by additional time between starting and running introduced by the provisioner lifecycle steps.

albertshau commented:
i'll probably break this up into smaller PRs
